lib/logstash/filters/grok/timeout_enforcer.rb

```diff
@@ -0,0 +1,77 @@
+class LogStash::Filters::Grok::TimeoutEnforcer
+  def initialize(logger, timeout_nanos)
+    @logger = logger
+    @running = true
+    @timeout_nanos = timeout_nanos
+
+    # Stores running matches with their start time, this is used to cancel long running matches
+    # Is a map of Thread => start_time
+    @timer_mutex = Mutex.new
+    @threads_to_start_time = {}        
+  end
+
+  def grok_till_timeout(event, grok, field, value)
+    begin
+      thread = Thread.current
+      start_thread_groking(thread)
+      yield
+    rescue ::LogStash::Filters::Grok::TimeoutException => e
+      # These fields aren't present at the time the exception was raised
+      # so we add them here.
+      # We could store this metadata in the @threads_to_start_time hash
+      # but that'd come at a perf cost and this works just as well.
+      e.grok = grok
```

Comment:
@andrewvc can we add a metric counter to measure the number of `timeout_errors` here? Will be really useful. There is already a grok namespaced metric